cli.go

```diff
@@ -73,29 +74,34 @@ func init() {
 	Default.Name = os.Args[0]
 }
 
-// Dispatch is yet another entry point which returns error
-func (app *App) Dispatch(args []string) error {
-	if len(args) == 0 {
-		app.PrintUsage()
-		return ErrUsage
+const mainCmd = ""
+
+func resolveCmd(args []string) (string, []string) {
+	if len(args) == 0 || strings.HasPrefix(args[0], "-") {
+		return mainCmd, args
 	}
+	return args[0], args[1:]
+}
 
-	cmdName := args[0]
+// Dispatch is yet another entry point which returns error
+func (app *App) Dispatch(args []string) error {
+	cmdName, args := resolveCmd(args)
 	if cmd, ok := app.Commands[cmdName]; ok {
 		flags := flag.NewFlagSet(cmdName, app.FlagErrorHandling)
 		flags.Usage = func() {
 			fmt.Fprintln(app.ErrorWriter, cmd.Usage(flags))
+			if cmdName == mainCmd {
+				app.PrintUsage()
+			}
```

Comment:
Becomes odd if we add `flag.Parse(args)` inside `actionMain` in the example:

```
% ./_example -h
Usage: if no subcommands are specified, run it.
Usage: ./_example <command> [<args>]

Commands:
    [main]    generate example
    smile     show smile
    up        count up!
```

We should make `app.PrintUsage` go nicely with the main command.

There seem two options:

- Using documentation of the main command inside `app.PrintUsage`
- Change `flags.Usage` to call only `app.PrintUsage` if the command is the main command